[Orleans] Document cancellation support (#46500)
dotnet/docs: docs/orleans/grains/cancellation-tokens.md

### Cancellation batching

Orleans optimizes performance by batching cancellation requests when many cancellations occur simultaneously. This reduces network overhead and CPU load during high-cancellation scenarios, such as when a service shuts down or when many operations time out simultaneously.

### Callback execution context

Callbacks registered on a <xref:System.Threading.CancellationToken> within a grain execute on the grain's scheduler. This lets you safely perform operations in the context of the grain from cancellation callbacks.

```csharp
public async Task ExampleWithCancellationCallbackAsync(CancellationToken cancellationToken = default)
{
    // Register a callback that will run on the grain's scheduler
    cancellationToken.Register(() =>
    {
        // This runs on the grain's execution context
        // Safe to access grain state here
        logger.LogInformation("Operation was canceled for grain {GrainId}", this.GetPrimaryKey());
    });
    
    // Continue with work...
    await DoWorkAsync(cancellationToken);
}
```

## Behavioral notes and limitations

### Cooperative cancellation

Cancellation in Orleans is cooperative, which means:

> - The grain method must actively check the <xref:System.Threading.CancellationToken> and respond to cancellation requests
> - Simply passing a canceled token doesn't automatically stop a running operation
> - For best results, check cancellation at regular intervals during long-running work

Reviewer: ```suggestion
- The grain method must actively check the <xref:System.Threading.CancellationToken> and respond to cancellation requests.
- Simply passing a canceled token doesn't automatically stop a running operation.
- For best results, check cancellation at regular intervals during long-running work.
```

### Timeout integration

When `CancelRequestOnTimeout` is enabled (default), Orleans automatically sends cancellation signals for timed-out requests:

```csharp
// If this call times out, Orleans sends a cancellation signal to the grain
await grain.LongRunningOperationAsync(cancellationToken);
```

### Cross-silo cancellation

Cancellation works across silo boundaries in distributed Orleans clusters:

> - Client-to-grain calls can be canceled regardless of which silo hosts the grain
> - Grain-to-grain calls support cancellation even when grains are on different silos
> - Network partitions might delay cancellation signal delivery

Reviewer: ```suggestion
- Client-to-grain calls can be canceled regardless of which silo hosts the grain.
- Grain-to-grain calls support cancellation even when grains are on different silos.
- Network partitions might delay cancellation signal delivery.
```

    }
    catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
    {
        // Expected cancellation from our token - clean up and re-throw
        await CleanupAsync();
        throw;
    }
    catch (OperationCanceledException)
    {
        // Cancellation from a different token (unexpected) - treat as error
        logger.LogWarning("Received unexpected cancellation");
        throw;
    }
    catch (Exception ex)
    {
        // Unexpected error - try fallback if not canceled
        if (cancellationToken.IsCancellationRequested)
        {
            throw new OperationCanceledException(cancellationToken);
        }
        
        logger.LogWarning(ex, "Primary processing failed, trying fallback");
        return await ProcessFallbackAsync(data, cancellationToken);
    }
}
```

When handling `OperationCanceledException`, always check if the cancellation is from the expected token using the `when` clause. This pattern distinguishes between expected cancellation (from your operation's token) and unexpected cancellation (from other sources).

### IAsyncEnumerable&lt;T&gt; behavior

For streaming scenarios with `IAsyncEnumerable<T>`:

> - **Pre-enumeration cancellation**: If the token is canceled before enumeration starts, no items are yielded and an `OperationCanceledException` is thrown immediately
> - **Mid-enumeration cancellation**: If canceled during enumeration, the stream stops at the next cancellation check point (typically the next `yield` or `await` operation)
> - **Exception propagation**: An `OperationCanceledException` is thrown to the consumer when cancellation occurs
> - **Resource cleanup**: The async enumerator's `DisposeAsync()` method is called automatically by `await foreach` to clean up resources
> - **Partial results**: Items yielded before cancellation are preserved and delivered to the consumer

Reviewer: ```suggestion
- **Pre-enumeration cancellation**: If the token is canceled before enumeration starts, no items are yielded and an `OperationCanceledException` is thrown immediately.
- **Mid-enumeration cancellation**: If canceled during enumeration, the stream stops at the next cancellation check point (typically the next `yield` or `await` operation).
- **Exception propagation**: An `OperationCanceledException` is thrown to the consumer when cancellation occurs.
- **Resource cleanup**: The async enumerator's `DisposeAsync()` method is called automatically by `await foreach` to clean up resources.
- **Partial results**: Items yielded before cancellation are preserved and delivered to the consumer.
```

The effectiveness of cancellation depends on how frequently the async iterator checks the cancellation token:

```csharp
public async IAsyncEnumerable<DataPoint> StreamDataAsync(
    int count, 
    [EnumeratorCancellation] CancellationToken cancellationToken = default)
{
    for (int i = 0; i < count; i++)
    {
        // Good: Check before each item
        cancellationToken.ThrowIfCancellationRequested();
        
        var data = await ProcessItemAsync(i);
        
        // Good: Check after long-running operations
        cancellationToken.ThrowIfCancellationRequested();
        
        yield return data;
        
        // Good: Use cancellation-aware operations
        await Task.Delay(100, cancellationToken);
    }
}
```

## Legacy: GrainCancellationToken and GrainCancellationTokenSource

Prior to the direct support for `System.Threading.CancellationToken` in grain methods, Orleans provided a specific mechanism using `GrainCancellationToken` and `GrainCancellationTokenSource`. **This approach is now considered legacy, and `System.Threading.CancellationToken` is the recommended method for implementing cancellation in Orleans grains.**

> The information below is provided for context on older systems or for specific backward compatibility needs. For new development, please refer to the <xref:System.Threading.CancellationToken> usage described earlier in this article.

Reviewer: ```suggestion
The following information is provided for context on older systems or for specific backward compatibility needs. For new development, see the <xref:System.Threading.CancellationToken> usage described earlier in this article.
```
